source/core/slang-hash.h

```diff
@@ -7,56 +7,102 @@
 
 namespace Slang
 {
-    typedef int HashCode;
+    // Ideally Hash codes should be unsigned types - makes accumulation simpler (as overflow/underflow behavior are defined)
+    // Only downside is around multiply, where unsigned multiply can be slightly slower on some targets.
 
-	inline int GetHashCode(double key)
+    // HashCode - size may vary by platform. Typically has 'best' combination of bits/performance. Should not be exposed externally as value from same input may change depending on compilation platform.
+    typedef unsigned int HashCode;
```

Comment:
I strongly agree with the move to make `HashCode` be unsigned.

Is there a strong argument for moving to a `HashCode` that is always 64 bits, or always pointer-sized, or do we think that sticking with the "natural" `int` size is always going to be the right choice?

Reply:
When performance is key and you support 32 and 64 word targets, having the 'normal' hash type be pointer sized makes a lot of sense. 

It could be now argued that the only targets we really care about are 64 bit. That making HashCode always 64 bits might diminish perf on 32 bit, and perhaps significantly (as in a few percent of overall execution), but if 32 bit targets are not a priority, then perhaps thats ok in the context of simplicity.

Reply:
I guess the central question for me is whether sticking with an `int`-sized hash value is a kind of "worst of both worlds" thing, in that there are 64-bit platforms where it will end up being 32 bits and others where it will be 64 bits, and that results in three different cases of hashing behavior that we need to think about for stuff like `Dictionary` and `HashSet` (ILP32, LP64, and ILP64).

Reply:
Yes - unsigned int isn't arguably the best choice. That this PR was in part about making it so you could easily have the option of changing the size. 

If you want a combination of perf/most bits across targets (that include 32/64 bits) it should probly be ptr sized. If you want some simplicity at the cost of perf on 32 bit, an argument can be made for it being 64 bit. 

To see what the impact should now be pretty easy to check.